pkg/util/system/cgroup_resource.go

```diff
@@ -18,124 +18,290 @@ package system
+	MemoryStatV2  = DefaultFactory.NewV2(MemoryStatName, MemoryStatName).WithCheckSupported(SupportedIfCgroupsV2)
+	MemoryMinV2   = DefaultFactory.NewV2(MemoryMinName, MemoryMinName).WithCheckSupported(SupportedIfCgroupsV2)
+	MemoryLowV2   = DefaultFactory.NewV2(MemoryLowName, MemoryLowName).WithCheckSupported(SupportedIfCgroupsV2)
+	MemoryHighV2  = DefaultFactory.NewV2(MemoryHighName, MemoryHighName).WithCheckSupported(SupportedIfCgroupsV2)
+
+	knownCgroupV2Resources = []Resource{
+		CPUQuotaV2,
+		CPUPeriodV2,
+		CPUStatV2,
+		MemoryLimitV2,
+		MemoryUsageV2,
+		MemoryStatV2,
+		MemoryMinV2,
+		MemoryLowV2,
+		MemoryHighV2,
+	}
 )
 
-type CgroupFile struct {
-	ResourceFileName string
-	Subfs            string
-	IsAnolisOS       bool
-	Validator        Validate
+var _ Resource = &CgroupResource{}
+
+type CgroupResource struct {
+	Type           ResourceType
+	FileName       string
+	Subfs          string
+	Supported      *bool
+	SupportMsg     string
+	CheckSupported func() (bool, string)
+	Validator      ResourceValidator
 }
 
-func ValidateCgroupValue(value *int64, parentDir string, file CgroupFile) bool {
-	if value == nil {
-		klog.V(5).Infof("validate fail, dir:%s, file:%s, value is nil!", parentDir, file.ResourceFileName)
-		return false
+func (c *CgroupResource) ResourceType() ResourceType {
+	if len(c.Type) > 0 {
+		return c.Type
 	}
-	if file.Validator != nil {
-		valid, msg := file.Validator.Validate(value)
-		if !valid {
-			klog.Warningf("validate fail! dir:%s, msg:%s", parentDir, msg)
+	return GetDefaultResourceType(c.Subfs, c.FileName)
+}
+
+func (c *CgroupResource) Path(parentDir string) string {
+	// get cgroup path
+	return filepath.Join(Conf.CgroupRootDir, c.Subfs, parentDir, c.FileName)
+}
+
+func (c *CgroupResource) IsSupported() (bool, string) {
+	if c.Supported == nil {
+		if c.CheckSupported == nil {
+			return false, "unknown support status"
 		}
-		return valid
+		supported, msg := c.CheckSupported()
+		c.Supported = &supported
+		c.SupportMsg = msg
+	}
+	return *c.Supported, c.SupportMsg
+}
+
+func (c *CgroupResource) IsValid(v string) (bool, string) {
+	if c.Validator == nil {
+		return true, ""
 	}
-	return true
+	return c.Validator.Validate(v)
+}
+
+func (c *CgroupResource) WithValidator(validator ResourceValidator) Resource {
+	c.Validator = validator
+	return c
+}
+
+func (c *CgroupResource) WithCheckSupported(checkSupportedFn func() (bool, string)) Resource {
+	c.CheckSupported = checkSupportedFn
+	return c
+}
+
+type CgroupResourceFactory interface {
+	New(filename string, subfs string) Resource // cgroup-v1 filename represents the resource type
+	NewV2(t ResourceType, filename string) Resource
+}
+
+type cgroupResourceFactoryImpl struct{}
+
+func NewCgroupResourceFactory() CgroupResourceFactory {
+	return &cgroupResourceFactoryImpl{}
+}
+
+func (f *cgroupResourceFactoryImpl) New(filename string, subfs string) Resource {
```

Comment:
will there be lots of if-else in `executor` since the same ResourceType(e.g. cfs quota) are divided to two files

Reply:
ResourceType defines the fine-grained cgroup resource.  `Executor` will pick the correct CgroupResource according to the ResourceType.
e.g. In cgroups-v1, CFSQuota, CFSPeriod refer to file `cpu.cfs_quota_us` and `cpu.cfs_period_us` respectively, while both of them refer to `cpu.max` in v2.

Reply:
It is convenient for updating cgroups with almost the same writing pattern on v1 and v2 (e.g. `memory.min`). It does require some if-else for cgroups with different patterns (e.g. `cpu.cfs_period_us`).